src/dmqmc_estimators.F90

```diff
@@ -1414,6 +1414,104 @@ subroutine calculate_rdm_renyi_2(subsys_info, rdm_lists, accumulated_probs_old,
 
     end subroutine calculate_rdm_renyi_2
 
+    subroutine calculate_green_function(sys, psip_list, weighted_sampling, green)
+
+        ! Calculate 2-RDM (Green Function) for the current density matrix.
+
+        ! In:
+        !   sys: system being studied.
+        !   psip_list: arrays containing information about psips.
+        !   weighted_sampling: information about weighted DMQMC sampling.
+        ! In/Out:
+        !   green: the 2-RDM matrix.
+
+        use system, only: sys_t
+        use qmc_data, only: particle_t
+        use determinants, only: det_info_t, alloc_det_info_t
+        use dmqmc_data, only: dmqmc_weighted_sampling_t
+        use excitations, only: excit_t, get_excitation_level
+        use proc_pointers, only: decoder_ptr
+        use replica_rdm, only: update_rdm
+        
+        type(sys_t), intent(in) :: sys
+        type(particle_t), intent(in), target :: psip_list
+        type(dmqmc_weighted_sampling_t), intent(in) :: weighted_sampling
+        real(p), intent(inout) :: green(:, :)
+
+        type(det_info_t) :: cdet1, cdet2
+        real(p) :: unweighted_walker_pop(psip_list%nspaces)
+        integer :: idet, nexcit
+
+        green = 0._p
+        call alloc_det_info_t(sys, cdet1, .false.)
+        call alloc_det_info_t(sys, cdet2, .false.)
+
+        do idet = 1, psip_list%nstates
+
+            ! Decode the two Slater determinants into cdet1 and cdet2.
+            cdet1%f => psip_list%states(:sys%basis%tot_string_len,idet)
+            cdet2%f => psip_list%states((sys%basis%tot_string_len+1):(2*sys%basis%tot_string_len),idet)
+            call decoder_ptr(sys, cdet1%f, cdet1)
+            call decoder_ptr(sys, cdet2%f, cdet2)
+
+            ! Need to get unweighted walker population for weighted sampling first.
+            ! [fixme] Excitation-related routines are called twice (once here and once in update_rdm). Maybe the redundant call 
+            !         can be removed by modifying a little the update_rdm procedure.
+            nexcit = get_excitation_level(cdet1%f, cdet2%f)
+            if (nexcit < 3) then
+                unweighted_walker_pop = real(psip_list%pops(:,idet),p)*weighted_sampling%probs(nexcit)/psip_list%pop_real_factor
```

Comment:
Why is weighted_sampling multiplying the walker's weight here? I though this was taken care of  in update_dmqmc_estimators?